spl/token/instructions.py

```diff
@@ -0,0 +1,804 @@
+    """The mint authority."""
+    amount: int
+    """Amount to mint."""
+    decimals: int
+    """Amount decimals."""
+    signers: List[PublicKey] = []
+    """Signing accounts if `mint_authority` is a multiSig."""
+
+
+class Burn2Params(NamedTuple):
+    """Burn2 token transaction params."""
+
+    program_id: PublicKey
+    """SPL Token program account."""
+    mint: PublicKey
+    """Public key of the minter account."""
+    account: PublicKey
+    """Account to burn tokens from."""
+    owner: PublicKey
+    """Owner of the account."""
+    amount: int
+    """Amount to burn."""
+    decimals: int
+    """Amount decimals."""
+    signers: List[PublicKey] = []
+    """Signing accounts if `owner` is a multiSig"""
+
+
+def decode_initialize_mint(instruction: TransactionInstruction) -> InitializeMintParams:
+    """Decode an initialize mint token instruction and retrieve the instruction params."""
+    validate_instruction_keys(instruction, 2)
+
+    parsed_data = INSTRUCTIONS_LAYOUT.parse(instruction.data)
+    validate_instruction_type(parsed_data, InstructionType.InitializeMint)
+
+    return InitializeMintParams(
+        decimals=parsed_data.args.decimals,
+        program_id=instruction.program_id,
+        mint=instruction.keys[0].pubkey,
+        mint_authority=PublicKey(parsed_data.args.mint_authority),
+        freeze_authority=PublicKey(parsed_data.args.freeze_authority)
+        if parsed_data.args.freeze_authority_option
+        else None,
+    )
+
+
+def decode_initialize_account(instruction: TransactionInstruction) -> InitializeAccountParams:
+    """Decode an initialize account token instruction and retrieve the instruction params."""
+    validate_instruction_keys(instruction, 4)
+
+    parsed_data = INSTRUCTIONS_LAYOUT.parse(instruction.data)
+    validate_instruction_type(parsed_data, InstructionType.InitializeAccount)
+
+    return InitializeAccountParams(
+        program_id=instruction.program_id,
+        account=instruction.keys[0].pubkey,
+        mint=instruction.keys[1].pubkey,
+        owner=instruction.keys[2].pubkey,
+    )
+
+
+def decode_initialize_multisig(instruction: TransactionInstruction) -> InitializeMultisigParams:
+    """Decode an initialize multisig account token instruction and retrieve the instruction params."""
+    parsed_data = INSTRUCTIONS_LAYOUT.parse(instruction.data)
+    validate_instruction_type(parsed_data, InstructionType.InitializeMultisig)
+    num_signers = parsed_data.args.m
+    validate_instruction_keys(instruction, 2 + num_signers)
+
+    return InitializeMultisigParams(
+        program_id=instruction.program_id,
+        multisig=instruction.keys[0].pubkey,
+        signers=[signer.pubkey for signer in instruction.keys[-num_signers:]],
+        m=num_signers,
+    )
+
+
+def decode_transfer(instruction: TransactionInstruction) -> TransferParams:
+    """Decode a transfer token transaction and retrieve the instruction params."""
+    validate_instruction_keys(instruction, 3)
```

Comment:
I am thinking if it will be nicer if we have a centralized, e.g. a `dict` which stores the minimum number of key length that is required for each instruction.

Reply:
Once we do that we can probably also extract between line 341, 343 and 344 to a generic function which takes two arguments: `instruction` and `instruction_type` and it will do the length check, parsing and then instruction type check.

Reply:
I was kind of lazy and did not create the mapping. But I did reduce some boilerplate. I think we'll end up redoing a lot of this anyways after Solana adds better ABI support.